racket/collects/xml/private/core.rkt

```diff
@@ -1,28 +1,86 @@
 #lang racket/base
-(require racket/contract/base)
+(require racket/runtime-path
+         (for-syntax racket/base))
 
 ;; Core structures needed for `xml/xexpr'
 
-(provide (all-defined-out))
+(provide permissive-xexprs
+         (struct-out source)
+         (struct-out comment)
+         (struct-out p-i)
+         (struct-out cdata)
+         valid-char?)
 
 ; permissive-xexprs : parameter bool
 (define permissive-xexprs (make-parameter #f))
 
+; support for serialization with contracts enforced on deserialization
+(module serialization-support racket/base
+  (require racket/serialize-structs
+           (for-syntax racket/syntax
+                       racket/struct-info
+                       racket/base))
+  (define ((cycles-not-allowed name))
+    (error name "invalid serialization;\n cycles not allowed"))
+  (define-for-syntax ((make-define-serializable-struct deserialize-info-mpi
+                                                       get-checked-constructor)
+                      stx)
+    (syntax-case stx ()
+      [(_ name-maybe-super (fld ...))
+       (with-syntax* ([(name super-fld-ref ...)
+                       (syntax-case #'name-maybe-super ()
+                         [name
+                          (identifier? #'name)
+                          #'(name)]
+                         [(name super)
+                          #`(name
+                             #,@(reverse (list-ref (extract-struct-info (syntax-local-value #'super))
+                                                   3)))])]
+                      [(name-fld-ref ...)
+                       (for/list ([id (in-list (syntax->list #'(fld ...)))])
+                         (format-id id "~a-~a" #'name id))]
+                      [deserialize-info:name-v0
+                       (format-id #'name "deserialize-info:~a-v0" #'name)])
+         #`(begin
+             (define-struct name-maybe-super (fld ...)
+               #:property prop:serializable
+               (make-serialize-info (λ (this)
+                                      (vector (super-fld-ref this) ...
+                                              (name-fld-ref this) ...))
+                                    (cons 'deserialize-info:name-v0 #,deserialize-info-mpi)
+                                    #f
+                                    (or (current-load-relative-directory) (current-directory)))
+               #:transparent)
+             (module+ deserialize-info
+               (provide deserialize-info:name-v0)
+               (define deserialize-info:name-v0
+                 (make-deserialize-info (#,get-checked-constructor name)
+                                        (cycles-not-allowed 'name))))))]))
+  (provide (for-syntax make-define-serializable-struct)))
+(require 'serialization-support)
+(define-runtime-module-path-index deserialize-info-mpi '(submod "." deserialize-info))
+(module+ deserialize-info
+  (define-runtime-module-path-index mpi-for-contracts "structures.rkt")
```

Comment:
I wanted to double-check that this doesn't need to be:
```racket
(define-runtime-module-path mpi-for-contracts xml/private/structures)
```

Reply:
Oh, I was thinking about the one that is currently `(define-runtime-module-path-index mpi-for-contracts "structures.rkt")`. I vaguely remembered that some ways of mixing styles of references could cause trouble for `raco exe` in some cases, but I've never actually encountered those problems myself, so I'm not very clear about it.

Reply:
I guess if someone referred to "core.rkt" through a file path instead of a module path, then there would be a difference for both of those relative module references, and there could be a mismatch. For that case, using relative paths as here might be better, or it might be worse, depending on the context. In any case, referencing "core.rkt" both through a file path and a module path would create other (related) problems with `raco exe`, and I don't think we need to worry about it.